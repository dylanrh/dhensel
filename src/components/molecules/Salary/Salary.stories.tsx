/* eslint-disable import/no-anonymous-default-export */
import { Salary, SalaryProps } from './Salary';

export default {
	title: 'molecules/Salary',
	component: Salary,
};

export const SalaryStory = (args: SalaryProps) => <Salary {...args} />;

SalaryStory.storyName = 'Salary';
SalaryStory.args = {
	salaryRange: {
		minimum: 100000,
		median: 150000,
		maximum: 200000,
	},
};
