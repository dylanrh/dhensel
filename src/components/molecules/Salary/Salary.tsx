'use client';

import { useEffect, useState } from 'react';

export interface SalaryProps {
	salaryRange: {
		minimum: number;
		median: number;
		maximum: number;
	};
}

const Salary = ({ salaryRange }: SalaryProps) => {
	const { median } = salaryRange;

	const roundNum = (num: number) => {
		return Math.round(num / 5000) * 5000;
	};

	const minVisible = roundNum(median * 0.5);
	const maxVisible = roundNum(median * 1.6);
	const minSad = roundNum(median * 0.6);
	const stillSad = roundNum(median * 0.7);
	const minAcceptable = roundNum(median * 0.8);
	const veryHappy = roundNum(median * 1.2);
	const overTheMoon = roundNum(median * 1.4);

	const [selectedSalary, setSelectedSalary] = useState(minAcceptable);
	const [emoji, setEmoji] = useState('❓');
	const [title, setTitle] = useState('❓');

	useEffect(() => {
		if (selectedSalary < minSad) {
			setEmoji('🤬');
			setTitle('Are you serious?');
		} else if (selectedSalary >= minSad && selectedSalary < stillSad) {
			setEmoji('😭');
			setTitle('Way too low');
		} else if (selectedSalary >= stillSad && selectedSalary < minAcceptable) {
			setEmoji('😢');
			setTitle('Too low');
		} else if (selectedSalary >= minAcceptable && selectedSalary < median) {
			setEmoji('😏');
			setTitle('getting there');
		} else if (selectedSalary >= median && selectedSalary < veryHappy) {
			setEmoji('🙂');
			setTitle('pretty good');
		} else if (selectedSalary >= veryHappy && selectedSalary < overTheMoon) {
			setEmoji('😀');
			setTitle('even better');
		} else if (selectedSalary >= overTheMoon && selectedSalary < maxVisible) {
			setEmoji('😃');
			setTitle('amazing');
		} else if (selectedSalary >= maxVisible) {
			setEmoji('🤑');
			setTitle('make it rain!');
		}
		// eslint-disable-next-line react-hooks/exhaustive-deps
	}, [selectedSalary]);

	return (
		<div>
			<p>
				<strong>Salary Expectation: </strong>
				There are several factors that influence the salary expectation I have
				for a future role, but here&lsquo;s a good approximation (please
				don&lsquo;t make me cry):
			</p>
			<input
				type="range"
				min={minVisible}
				max={maxVisible}
				value={selectedSalary}
				step={5_000}
				onChange={(e) => setSelectedSalary(Number(e.target.value))}
				className="mx-2"
				data-testid="salaryRange"
			/>
			<label htmlFor="salaryRange">
				{Intl.NumberFormat('en-GB', {
					style: 'currency',
					currency: 'GBP',
				}).format(selectedSalary)}
			</label>
			<p>
				<strong>Happiness Score:</strong>
				<span className="text-3xl" data-testid="emoji" title={title}>
					{` ${emoji}`}
				</span>
			</p>
		</div>
	);
};

export { Salary };
