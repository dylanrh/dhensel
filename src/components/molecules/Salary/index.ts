'use client';

export * from './Salary';
