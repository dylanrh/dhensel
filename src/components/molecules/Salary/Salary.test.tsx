import { Salary } from '../Salary';

import { fireEvent, render, waitFor } from '@testing-library/react';

describe('Salary', () => {
	const testRange = {
		minimum: 100000,
		median: 150000,
		maximum: 200000,
	};

	it('renders correctly', () => {
		const { container } = render(<Salary salaryRange={testRange} />);
		expect(container).toMatchSnapshot();
	});

	it('switches the emoji when changing the range', async () => {
		const { getByTestId } = render(<Salary salaryRange={testRange} />);
		const rangeInput = getByTestId('salaryRange');
		const emoji = getByTestId('emoji');
		expect(emoji).toHaveAccessibleDescription('getting there');

		fireEvent.change(rangeInput, { target: { value: '70000' } });
		await waitFor(() => {
			expect(emoji).toHaveAccessibleDescription(/are you serious?/i);
		});

		fireEvent.change(rangeInput, { target: { value: '90000' } });
		expect(emoji).toHaveAccessibleDescription(/way too low/i);

		fireEvent.change(rangeInput, { target: { value: '95000' } });
		expect(emoji).toHaveAccessibleDescription(/too low/i);
	});
});
