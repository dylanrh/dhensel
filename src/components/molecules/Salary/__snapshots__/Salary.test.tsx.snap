// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Salary renders correctly 1`] = `
<div>
  <div>
    <p>
      <strong>
        Salary Expectation: 
      </strong>
      There are several factors that influence the salary expectation I have for a future role, but here‘s a good approximation (please don‘t make me cry):
    </p>
    <input
      class="mx-2"
      data-testid="salaryRange"
      max="240000"
      min="75000"
      step="5000"
      type="range"
      value="120000"
    />
    <label
      for="salaryRange"
    >
      £120,000.00
    </label>
    <p>
      <strong>
        Happiness Score:
      </strong>
      <span
        class="text-3xl"
        data-testid="emoji"
        title="getting there"
      >
         😏
      </span>
    </p>
  </div>
</div>
`;
